File: golang_port/core/impl/services/translate_executor.go
// Package services holds core use-case orchestrators. This file ports
// crates/gateway-http-anthropic/src/translate_executor.rs:78-237: the
// local-command status executor that produces a Gateway-owned status
// document for translated Claude Code slash commands (currently "/status"),
// with non-blocking live usage enrichment and rate-limit/spend-control
// normalization.
package services

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/Bytonomics/cld-gateway/core"
	"github.com/Bytonomics/cld-gateway/netpolicy"
)

// SessionInfo carries Gateway's local session/thread information for status
// display, mirroring translate_executor.rs SessionInfo (:28-33).
type SessionInfo struct {
	ThreadID       *string
	ThreadName     *string
	AccountDisplay *string
}

// ExecutorRuntime is the runtime context passed to executor functions,
// mirroring translate_executor.rs ExecutorRuntime (:8-25). It intentionally
// depends only on core.Secret/netpolicy.Client rather than the codex backend
// client package, keeping this file free of a codex-specific import.
type ExecutorRuntime struct {
	HasCredentials  bool
	AccessToken     core.Secret
	AccountID       string
	BaseURL         string
	HTTPClient      *netpolicy.Client
	CurrentModel    *string
	SessionInfo     SessionInfo
	GatewayVersion  string
	ConfigPath      *string
	ResolvedModel   *string
	CurrentDir      *string
	ReasoningEffort *string
}

// PostResultFn is the post-result wrapper function signature: takes executor
// JSON and packaged command body, returns final output text. Mirrors
// PostResultFn (:36).
type PostResultFn func(executorJSON map[string]any, packagedBody string) string

// CommandExecutorNames is the registry of translated command names that have
// executor functions, mirroring COMMAND_EXECUTOR_NAMES (:40).
var CommandExecutorNames = []string{"status"}

// CommandPostResults maps normalized command name to its post-result wrapper
// function, mirroring COMMAND_POST_RESULTS (:45-46).
var CommandPostResults = map[string]PostResultFn{
	"status": PostResultForTranslatedCommand,
}

// normalizeCommandName strips leading/trailing whitespace and a leading
// slash, matching the Rust normalization at translate_executor.rs:62.
func normalizeCommandName(name string) string {
	return strings.TrimPrefix(strings.TrimSpace(name), "/")
}

// ExecuteTranslatedCommand executes a translated command if one is present,
// mirroring execute_translated_command (:53-76). commandName == nil mirrors
// Option::None. Returns (nil, nil) when no translated command was detected;
// an error return is reserved for a registered command whose execution
// fails explicitly (none do today, matching the Rust source).
func ExecuteTranslatedCommand(ctx context.Context, commandName *string, runtime *ExecutorRuntime) (map[string]any, error) {
	if commandName == nil {
		return nil, nil
	}

	normalized := normalizeCommandName(*commandName)

	found := false
	for _, n := range CommandExecutorNames {
		if n == normalized {
			found = true
			break
		}
	}
	if !found {
		return nil, nil
	}

	switch normalized {
	case "status":
		return executeStatusCommand(ctx, runtime), nil
	default:
		return nil, nil
	}
}

// executeStatusCommand builds a Gateway-owned status document with local
// session info and optional usage data, mirroring execute_status_command
// (:81-146). It returns immediately with local state; usage enrichment is
// best-effort and never fails the executor.
func executeStatusCommand(ctx context.Context, runtime *ExecutorRuntime) map[string]any {
	baseURL := strings.TrimRight(runtime.BaseURL, "/")

	accountID := "unavailable"
	if runtime.HasCredentials {
		accountID = runtime.AccountID
	}

	timestamp := time.Now().Unix()

	status := map[string]any{
		"status_type":  "gateway_status",
		"generated_at": timestamp,
		"gateway": map[string]any{
			"version":     runtime.GatewayVersion,
			"config_path": derefOrNil(runtime.ConfigPath),
			"current_dir": derefOrNil(runtime.CurrentDir),
		},
		"session": map[string]any{
			"thread_id":       derefOrNil(runtime.SessionInfo.ThreadID),
			"thread_name":     derefOrNil(runtime.SessionInfo.ThreadName),
			"account_display": derefOrNil(runtime.SessionInfo.AccountDisplay),
		},
		"model": map[string]any{
			"requested":        derefOrNil(runtime.CurrentModel),
			"resolved":         resolvedModelOrCurrent(runtime),
			"reasoning_effort": derefOrNil(runtime.ReasoningEffort),
		},
		"provider": map[string]any{
			"base_url": baseURL,
		},
		"auth": map[string]any{
			"account_id": accountID,
		},
		"usage_state":   "pending",
		"plan_type":     nil,
		"rate_limits":   nil,
		"spend_control": nil,
		"usage_raw":     nil,
	}

	// Errors are captured in the status document, not escalated as executor
	// failure (:126).
	if usageData, err := fetchLiveUsageData(ctx, runtime); err == nil {
		if planType, ok := usageData["plan_type"]; ok {
			status["plan_type"] = planType
		}
		status["rate_limits"] = normalizeRateLimits(usageData)
		if spend, ok := usageData["spend_control"].(map[string]any); ok {
			status["spend_control"] = normalizeSpendControl(spend)
		}
		status["usage_raw"] = usageData
		status["usage_state"] = "current"
	} else {
		status["usage_state"] = "stale_or_unavailable"
	}

	return status
}

// fetchLiveUsageData fetches live usage/rate-limit data from the upstream
// Codex API, mirroring fetch_live_usage_data (:150-183).
func fetchLiveUsageData(ctx context.Context, runtime *ExecutorRuntime) (map[string]any, error) {
	if !runtime.HasCredentials {
		return nil, errors.New("credentials_unavailable")
	}

	baseURL := strings.TrimRight(runtime.BaseURL, "/")
	url := baseURL + "/api/codex/usage"

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("usage_fetch_policy_error: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+runtime.AccessToken.Expose())
	req.Header.Set("chatgpt-account-id", runtime.AccountID)

	httpClient := runtime.HTTPClient
	if httpClient == nil {
		httpClient = netpolicy.New(nil)
	}

	res, err := httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("usage_fetch_transport_error: %w", err)
	}
	defer func() { _ = res.Body.Close() }()

	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return nil, fmt.Errorf("usage_fetch_status_%d", res.StatusCode)
	}

	body, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, fmt.Errorf("usage_fetch_body_error: %w", err)
	}

	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		return nil, fmt.Errorf("usage_fetch_parse_error: %w", err)
	}
	return parsed, nil
}

// normalizeRateLimits normalizes rate-limit data from the upstream usage
// response into stable Gateway-owned fields, mirroring normalize_rate_limits
// (:186-237).
func normalizeRateLimits(usage map[string]any) map[string]any {
	limits := map[string]any{}

	if rl, ok := usage["rate_limit"].(map[string]any); ok {
		primary := map[string]any{
			"allowed":       rl["allowed"],
			"limit_reached": rl["limit_reached"],
		}
		if pw, ok := rl["primary_window"].(map[string]any); ok {
			primary["used_percent"] = pw["used_percent"]
			primary["reset_at"] = pw["reset_at"]
			primary["window_seconds"] = pw["limit_window_seconds"]
		}
		limits["primary"] = primary

		if sw, ok := rl["secondary_window"].(map[string]any); ok {
			limits["secondary"] = map[string]any{
				"used_percent":   sw["used_percent"],
				"reset_at":       sw["reset_at"],
				"window_seconds": sw["limit_window_seconds"],
			}
		}
	}

	if additional, ok := usage["additional_rate_limits"].([]any); ok {
		entries := make([]any, 0, len(additional))
		for _, raw := range additional {
			entry, ok := raw.(map[string]any)
			if !ok {
				continue
			}
			name, ok := entry["limit_name"].(string)
			if !ok {
				continue
			}
			rl, ok := entry["rate_limit"].(map[string]any)
			if !ok {
				continue
			}
			pw, ok := rl["primary_window"].(map[string]any)
			if !ok {
				continue
			}
			entries = append(entries, map[string]any{
				"limit_name":     name,
				"allowed":        rl["allowed"],
				"limit_reached":  rl["limit_reached"],
				"used_percent":   pw["used_percent"],
				"reset_at":       pw["reset_at"],
				"window_seconds": pw["limit_window_seconds"],
			})
		}
		if len(entries) > 0 {
			limits["additional"] = entries
		}
	}

	return limits
}

// normalizeSpendControl normalizes spend-control data from the upstream
// usage response, mirroring normalize_spend_control (:240-256).
func normalizeSpendControl(spend map[string]any) map[string]any {
	result := map[string]any{
		"reached": spend["reached"],
	}
	if limit, ok := spend["individual_limit"].(map[string]any); ok {
		result["individual_limit"] = map[string]any{
			"source":            limit["source"],
			"limit":             limit["limit"],
			"used":              limit["used"],
			"remaining":         limit["remaining"],
			"used_percent":      limit["used_percent"],
			"remaining_percent": limit["remaining_percent"],
			"reset_at":          limit["reset_at"],
		}
	}
	return result
}

// PostResultForTranslatedCommand wraps executor JSON result with packaged
// command instructions if provided, mirroring
// post_result_for_translated_command (:258-282). An empty packaged body
// (after trimming) yields JSON-only output, so empty packaged prompts never
// inject synthetic scaffolding.
func PostResultForTranslatedCommand(executorJSON map[string]any, packagedBody string) string {
	packagedTrimmed := strings.TrimSpace(packagedBody)

	jsonBytes, err := json.MarshalIndent(executorJSON, "", "  ")
	jsonStr := string(jsonBytes)
	if err != nil {
		jsonStr = fmt.Sprintf("%v", executorJSON)
	}

	if packagedTrimmed == "" {
		return jsonStr
	}
	return packagedTrimmed + "\n\n" + jsonStr
}

// GetPostResultFunction returns the post-result wrapper function for a
// translated command, if registered, mirroring get_post_result_function
// (:285-291).
func GetPostResultFunction(commandName string) (PostResultFn, bool) {
	normalized := normalizeCommandName(commandName)
	fn, ok := CommandPostResults[normalized]
	return fn, ok
}

func derefOrNil(s *string) any {
	if s == nil {
		return nil
	}
	return *s
}

func resolvedModelOrCurrent(runtime *ExecutorRuntime) any {
	if runtime.ResolvedModel != nil {
		return *runtime.ResolvedModel
	}
	if runtime.CurrentModel != nil {
		return *runtime.CurrentModel
	}
	return nil
}

File: old_rust/crates/gateway-http-anthropic/src/translate_executor.rs
#![forbid(unsafe_code)]

use gateway_auth_codex::CodexCredentials;
use gateway_backend_codex::client::CodexBackendClient;
use serde_json::json;

/// Runtime context passed to executor functions.
pub struct ExecutorRuntime {
    pub credentials: Option<CodexCredentials>,
    pub backend_client: CodexBackendClient,
    /// Current model being used in this session
    pub current_model: Option<String>,
    /// Gateway's local session/thread information
    pub session_info: SessionInfo,
    /// Gateway binary version (from `CARGO_PKG_VERSION`)
    pub gateway_version: &'static str,
    /// Path to the gateway config file being used
    pub config_path: Option<String>,
    /// The resolved model name after model resolution
    pub resolved_model: Option<String>,
    /// Current working directory of the gateway process
    pub current_dir: Option<String>,
    /// Reasoning effort from request (if present in client metadata)
    pub reasoning_effort: Option<String>,
}

/// Session information for status display
#[derive(Debug, Clone)]
pub struct SessionInfo {
    pub thread_id: Option<String>,
    pub thread_name: Option<String>,
    pub account_display: Option<String>,
}

/// Post-result wrapper function signature: takes executor JSON and packaged body, returns final output
type PostResultFn = fn(&serde_json::Value, &str) -> String;

/// Registry of translated command names that have executor functions.
/// Allows future translated commands to be registered without code branching changes.
static COMMAND_EXECUTOR_NAMES: &[&str] = &["status"];

/// Registry of translated commands and their post-result wrapper functions.
/// Maps normalized command name to post-result function.
/// Allows future translated commands to customize how executor output is wrapped with packaged prompt text.
static COMMAND_POST_RESULTS: &[(&str, PostResultFn)] =
    &[("status", post_result_for_translated_command)];

/// Executes a translated command if one is present in the metadata.
///
/// Returns the executor JSON result if a translated command was found and executed.
/// Returns None if no translated command was detected.
/// Returns an error if the command was found but execution failed (explicit, not silent degradation).
pub async fn execute_translated_command(
    command_name: Option<&str>,
    runtime: &ExecutorRuntime,
) -> Result<Option<serde_json::Value>, String> {
    let Some(cmd) = command_name else {
        return Ok(None);
    };

    // Normalize: remove leading slash and whitespace.
    let normalized = cmd.trim().trim_start_matches('/');

    // Check if command exists in the registry
    if !COMMAND_EXECUTOR_NAMES.contains(&normalized) {
        return Ok(None);
    }

    // Dispatch to the appropriate async executor
    let result = match normalized {
        "status" => execute_status_command(runtime).await,
        _ => return Ok(None),
    };

    Ok(Some(result))
}

/// Executes the /status command.
/// Builds a Gateway-owned status document with local session info and optional usage data.
/// Returns immediately with local state; usage data enrichment is non-blocking.
async fn execute_status_command(runtime: &ExecutorRuntime) -> serde_json::Value {
    let base_url = runtime.backend_client.base_url();
    let account_id = runtime
        .credentials
        .as_ref()
        .map_or_else(|| "unavailable".to_string(), |c| c.account_id.clone());

    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map_or(0, |d| d.as_secs());

    // Build local status immediately (non-blocking).
    let mut status = json!({
        "status_type": "gateway_status",
        "generated_at": timestamp,
        "gateway": {
            "version": runtime.gateway_version,
            "config_path": runtime.config_path.as_deref(),
            "current_dir": runtime.current_dir.as_deref(),
        },
        "session": {
            "thread_id": runtime.session_info.thread_id.as_deref(),
            "thread_name": runtime.session_info.thread_name.as_deref(),
            "account_display": runtime.session_info.account_display.as_deref(),
        },
        "model": {
            "requested": runtime.current_model.as_deref(),
            "resolved": runtime.resolved_model.as_deref()
                .or(runtime.current_model.as_deref()),
            "reasoning_effort": runtime.reasoning_effort.as_deref(),
        },
        "provider": {
            "base_url": base_url.trim_end_matches('/'),
        },
        "auth": {
            "account_id": account_id,
        },
        "usage_state": "pending",
        "plan_type": null,
        "rate_limits": null,
        "spend_control": null,
        "usage_raw": null,
    });

    // Attempt to enrich with live usage data.
    // Errors are captured in the status document, not escalated as executor failure.
    if let Ok(usage_data) = fetch_live_usage_data(runtime).await {
        // Surface plan_type at top level
        if let Some(plan_type) = usage_data.get("plan_type") {
            status["plan_type"] = plan_type.clone();
        }
        // Normalize rate-limit summary
        status["rate_limits"] = normalize_rate_limits(&usage_data);
        // Surface spend-control summary
        if let Some(spend) = usage_data.get("spend_control") {
            status["spend_control"] = normalize_spend_control(spend);
        }
        // Keep raw blob for consumers that need full detail
        status["usage_raw"] = usage_data;
        status["usage_state"] = "current".into();
    } else {
        status["usage_state"] = "stale_or_unavailable".into();
    }

    status
}

/// Fetches live usage/rate-limit data from the upstream Codex API.
/// Errors are captured but do not cause the executor to fail.
async fn fetch_live_usage_data(runtime: &ExecutorRuntime) -> Result<serde_json::Value, String> {
    let Some(creds) = runtime.credentials.as_ref() else {
        return Err("credentials_unavailable".to_string());
    };

    let base_url = runtime.backend_client.base_url();
    let url = format!("{}/api/codex/usage", base_url.trim_end_matches('/'));
    let authorization = format!("Bearer {}", creds.access_token.expose());
    let account_id = &creds.account_id;

    let http_client = runtime.backend_client.http_client();

    let response = http_client
        .get(&url)
        .map_err(|err| format!("usage_fetch_policy_error: {err}"))?
        .header("Authorization", &authorization)
        .header("chatgpt-account-id", account_id)
        .execute()
        .await
        .map_err(|err| format!("usage_fetch_transport_error: {err}"))?;

    let status = response.status();
    if !status.is_success() {
        return Err(format!("usage_fetch_status_{}", status.as_u16()));
    }

    let body = response
        .text()
        .await
        .map_err(|err| format!("usage_fetch_body_error: {err}"))?;

    serde_json::from_str::<serde_json::Value>(&body)
        .map_err(|err| format!("usage_fetch_parse_error: {err}"))
}

/// Normalizes rate-limit data from the upstream usage response into stable Gateway-owned fields.
fn normalize_rate_limits(usage: &serde_json::Value) -> serde_json::Value {
    let mut limits = json!({});

    // Primary rate limit
    if let Some(rl) = usage.get("rate_limit") {
        limits["primary"] = json!({
            "allowed": rl.get("allowed"),
            "limit_reached": rl.get("limit_reached"),
        });
        if let Some(pw) = rl.get("primary_window") {
            limits["primary"]["used_percent"] = pw.get("used_percent").cloned().unwrap_or_default();
            limits["primary"]["reset_at"] = pw.get("reset_at").cloned().unwrap_or_default();
            limits["primary"]["window_seconds"] =
                pw.get("limit_window_seconds").cloned().unwrap_or_default();
        }
        if let Some(sw) = rl.get("secondary_window") {
            limits["secondary"] = json!({
                "used_percent": sw.get("used_percent"),
                "reset_at": sw.get("reset_at"),
                "window_seconds": sw.get("limit_window_seconds"),
            });
        }
    }

    // Additional per-limit entries
    if let Some(additional) = usage
        .get("additional_rate_limits")
        .and_then(|v| v.as_array())
    {
        let entries: Vec<serde_json::Value> = additional
            .iter()
            .filter_map(|entry| {
                let name = entry.get("limit_name")?.as_str()?;
                let rl = entry.get("rate_limit")?;
                let pw = rl.get("primary_window")?;
                Some(json!({
                    "limit_name": name,
                    "allowed": rl.get("allowed"),
                    "limit_reached": rl.get("limit_reached"),
                    "used_percent": pw.get("used_percent"),
                    "reset_at": pw.get("reset_at"),
                    "window_seconds": pw.get("limit_window_seconds"),
                }))
            })
            .collect();
        if !entries.is_empty() {
            limits["additional"] = serde_json::Value::Array(entries);
        }
    }

    limits
}

/// Normalizes spend-control data from the upstream usage response.
fn normalize_spend_control(spend: &serde_json::Value) -> serde_json::Value {
    let mut result = json!({
        "reached": spend.get("reached"),
    });
    if let Some(limit) = spend.get("individual_limit") {
        result["individual_limit"] = json!({
            "source": limit.get("source"),
            "limit": limit.get("limit"),
            "used": limit.get("used"),
            "remaining": limit.get("remaining"),
            "used_percent": limit.get("used_percent"),
            "remaining_percent": limit.get("remaining_percent"),
            "reset_at": limit.get("reset_at"),
        });
    }
    result
}

/// Wraps executor JSON result with packaged command instructions if provided.
///
/// If `packaged_body` is non-empty after trimming:
///   Returns: `"{packaged_body}\n\n{executor_json_pretty}"`
///
/// If `packaged_body` is empty:
///   Returns: `"{executor_json_pretty}"` only
///
/// This ensures empty packaged prompt does not inject synthetic scaffolding.
pub fn post_result_for_translated_command(
    executor_json: &serde_json::Value,
    packaged_body: &str,
) -> String {
    let packaged_trimmed = packaged_body.trim();
    let json_str =
        serde_json::to_string_pretty(executor_json).unwrap_or_else(|_| executor_json.to_string());

    if packaged_trimmed.is_empty() {
        // No packaged instructions, return JSON only (no synthetic scaffolding)
        json_str
    } else {
        // Prepend packaged instructions to JSON
        format!("{packaged_trimmed}\n\n{json_str}")
    }
}

/// Returns the post-result wrapper function for a translated command, if registered.
pub fn get_post_result_function(command_name: &str) -> Option<PostResultFn> {
    let normalized = command_name.trim().trim_start_matches('/');
    COMMAND_POST_RESULTS
        .iter()
        .find(|(name, _)| *name == normalized)
        .map(|(_, func)| *func)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_runtime(with_creds: bool) -> ExecutorRuntime {
        let creds = if with_creds {
            Some(gateway_auth_codex::CodexCredentials {
                access_token: gateway_core::Secret::new("test-token".to_string()),
                account_id: "test-account-123".to_string(),
            })
        } else {
            None
        };
        ExecutorRuntime {
            credentials: creds,
            backend_client: CodexBackendClient::default(),
            current_model: Some("claude-sonnet-4-20250514".to_string()),
            session_info: SessionInfo {
                thread_id: None,
                thread_name: None,
                account_display: Some("test-account-123".to_string()),
            },
            gateway_version: "0.0.0-test",
            config_path: Some("/tmp/test-config.yml".to_string()),
            resolved_model: Some("gpt-4o".to_string()),
            current_dir: Some("/tmp/test-workdir".to_string()),
            reasoning_effort: Some("high".to_string()),
        }
    }

    #[tokio::test]
    async fn execute_translated_command_dispatches_status() {
        let runtime = test_runtime(false);
        let result = execute_translated_command(Some("/status"), &runtime).await;
        let json = result.expect("no error").expect("executor found");
        assert_eq!(json["status_type"], "gateway_status");
    }

    #[tokio::test]
    async fn execute_translated_command_returns_none_for_unknown() {
        let runtime = test_runtime(false);
        let result = execute_translated_command(Some("/nonexistent"), &runtime).await;
        assert!(result.expect("no error").is_none());
    }

    #[tokio::test]
    async fn execute_translated_command_returns_none_for_no_command() {
        let runtime = test_runtime(false);
        let result = execute_translated_command(None, &runtime).await;
        assert!(result.expect("no error").is_none());
    }

    #[tokio::test]
    async fn status_executor_returns_structured_json_with_all_fields() {
        let runtime = test_runtime(false);
        let json = execute_status_command(&runtime).await;

        // Timestamp
        assert!(json["generated_at"].as_u64().unwrap() > 0);

        // Gateway fields
        assert_eq!(json["gateway"]["version"], "0.0.0-test");
        assert_eq!(json["gateway"]["config_path"], "/tmp/test-config.yml");
        assert_eq!(json["gateway"]["current_dir"], "/tmp/test-workdir");

        // Model fields
        assert_eq!(json["model"]["requested"], "claude-sonnet-4-20250514");
        assert_eq!(json["model"]["resolved"], "gpt-4o");
        assert_eq!(json["model"]["reasoning_effort"], "high");

        // Provider fields
        assert!(json["provider"]["base_url"].is_string());

        // Auth fields (no creds → "unavailable")
        assert_eq!(json["auth"]["account_id"], "unavailable");

        // Usage degraded without creds
        assert_eq!(json["usage_state"], "stale_or_unavailable");
        assert!(json["plan_type"].is_null());
        assert!(json["rate_limits"].is_null());
        assert!(json["spend_control"].is_null());
    }

    #[tokio::test]
    async fn status_executor_with_creds_shows_account_id() {
        let runtime = test_runtime(true);
        let json = execute_status_command(&runtime).await;
        assert_eq!(json["auth"]["account_id"], "test-account-123");
        // Usage fetch will fail (no real backend), but that's graceful degradation
        assert_eq!(json["usage_state"], "stale_or_unavailable");
    }

    #[test]
    fn post_result_with_empty_packaged_body_returns_json_only() {
        let json = serde_json::json!({"status_type": "gateway_status"});
        let result = post_result_for_translated_command(&json, "");
        // Should contain the JSON content but no prepended instructions
        assert!(result.contains("gateway_status"));
        assert!(!result.contains("Execute"));
        // Should NOT start with any instruction text — just JSON
        assert!(result.trim_start().starts_with('{'));
    }

    #[test]
    fn post_result_with_nonempty_packaged_body_prepends_instructions() {
        let json = serde_json::json!({"status_type": "gateway_status"});
        let result = post_result_for_translated_command(&json, "Show this status info clearly.");
        assert!(result.starts_with("Show this status info clearly."));
        assert!(result.contains("gateway_status"));
    }

    #[test]
    fn get_post_result_function_finds_status() {
        assert!(get_post_result_function("status").is_some());
        assert!(get_post_result_function("/status").is_some());
    }

    #[test]
    fn get_post_result_function_returns_none_for_unknown() {
        assert!(get_post_result_function("nonexistent").is_none());
    }

    #[test]
    fn missing_executor_for_classified_translate_is_distinguishable() {
        // When a command is classified Translate but has no executor,
        // execute_translated_command returns Ok(None). The handler in lib.rs
        // treats this as an explicit error for classified-Translate commands.
        // This test verifies the registry lookup behavior.
        assert!(COMMAND_EXECUTOR_NAMES.contains(&"status"));
        assert!(!COMMAND_EXECUTOR_NAMES.contains(&"nonexistent"));
    }
}
